### ISSUE TITLE
Parallelization for embarrassingly parallel tasks

### ISSUE THREAD

**Issue by michaelaye** [pydata/pandas#5751]
I would like to promote the idea of applying multiprocessing.Pool() execution of embarrassingly parallel tasks, e.g. the application of a function to a large number of columns.
Obviously, there's some overhead to set it up, so there will be a lower limit of number of columns from which on this only can be faster than the already fast cython approach. I will be adding my performance tests later to this issue.
I was emailing @jreback about this and he added the following remarks/complications:

* transferring numpy data is not that efficient - but still prob good enough
* you can transfer (pickle) lambda type functions so maybe need to look at using a library called dill ( which solves this problem) - possibly could slightly modify msgpack to do this though (and is already pretty efficient at transferring other types of objects)
* could also investigate joblib - I think statsmodels uses it [ed: That seems to be correct, I read in their group about joblib]
* I would create a new top level dir core/parallel for this type of stuff
* the strategy in this link could be a good way to follow: http://stackoverflow.com/questions/17785275/share-large-read-only-numpy-array-between-multiprocessing-processes 

links:
http://stackoverflow.com/questions/13065172/multiprocess-python-numpy-code-for-processing-data-faster
http://docs.cython.org/src/userguide/parallelism.html
http://distarray.readthedocs.org/en/v0.5.0/

**Comment by jseabold** [comment]
Linking this with #3202.

**Comment by abell25** [comment]
Has there been any progress on this?  I would like to help if we need developers to work on this!  A parallel datatable would be *really useful* considering I have a 36 core machine using only 1 thread :(  Pyspark has an parallel implementation but it is really slow on a single machine.

**Comment by shoyer** [comment]
There are a lot of interesting new tools in this space: dask, Ibis, SFrame. I would recommend taking a look at some of these. Extending pandas to natively support out of core computation would be great but it would be quite a lot of work and it's not on anyone's immediate agenda. Honestly it's probably best left to other projects.

**Comment by jreback** [comment]
using pandas in a parallel way requires a fair amount of work on the implementation. Fortunately [dask](https://dask.readthedocs.org/en/latest/) has already done this, and it works nicely with pandas (in fact, we released the gil and created the ``CategoricalIndex`` to facilitate this. So I would recommend this for embarrasingly parallel (and in fact generalized parallel computations).

I am going to change this issue to one of docs, as I think we should add a section to ``enhancingperf.rst`` to point the way to dask.

cc @mrocklin

**Comment by jreback** [comment]
Further, it may be desirable to provide an interface from the pandas side to ``dask`` (and possibly some of the out-of-core / parallel computation libraries), e.g. something along the lines of

``df.apply(f, engine='dask')``. Which can dispatch the computation. However, this would not be a generalized mechanism as often what you really want to do is do a series of computations. So not sure what / if anything we should do here.

**Comment by mrocklin** [comment]
I'd be quite happy to work with people who are interested in paralleliing pandas-like computations with dask.  I suspect that this will involve a bit of work both on ensuring that `dask.dataframe` encompasses all necessary Pandas API and in ensuring that Pandas releases the GIL in the appropriate locations.  

Making this work very well probably requires action by a few parties, a user with a driving application, a core pandas developer, and a core dask developer.  Jeff and I have had good interactions on this in the past; it'd be good to add in some people with concrete use cases.

**Comment by TomAugspurger** [comment]
Is the expectation that `df.apply(f, engine='dask')` (or any other engine) *must* return the same result as `df.apply(f)`? Or could dask return a dask graph? I'd be interested to see what returning a graph would allow you to do, and how hard it would be to implement? Would love to talk about this today @mrocklin

**Comment by jreback** [comment]
@TomAugspurger 

IF we did this, then I think this would simply do the computation (in parallel) and return the result. Kind of a hey, just use multi-cores type of thing. My viewpoint is that this gives the average pandas users a big hammer, so instead of vectorizing (or using cython/numba), they IMMEDIATELY go towards parallel when it is not necessary.

So that is my hesitation here. (others have argued that you simply document/educate and provide the tools). My argument is that maybe we should NOT do this in pandas AT ALL, and if you are sophisticated enough to use parallel execution, then using dask mainline (or ibis et al.) is the correct way to go about this.

**Comment by mrocklin** [comment]
My views align with @jreback 's here.  I think that if your data already lives in a pandas DataFrame then eager parallel execution probably provides most of the benefits that you should expect without anything unexpected.

**Comment by TomAugspurger** [comment]
Very reasonable.

**Comment by JanSchulz** [comment]
On the other hand, for novice pandas users using numba or even dask (and sometimes even vectorizing) is not so easy while getting a performance boost without needing to do anything (aka "use parallel everywhere" config option set to `True`) is a very nice benefit for everyone, even power users.

**Comment by jreback** [comment]
I think we could add an engine= argument

and allow say: numba, dask (if the are installed)

which gets most of the way there

**Comment by nehalecky** [comment]
Agree with the comments above :
>I think that if your data already lives in a pandas DataFrame then eager parallel execution probably provides most of the benefits that you should expect without anything unexpected.

That's dead on. 

Also, I very much like the idea of an engine = argument option. This would be a huge benefit for most end users, especially those using pandas as core dependency in their own applications—immediate parallelism across `.map` and `.apply` with the inclusions of a dependency (ala dask) and / or JIT (ala numba) and a simple configuration option. 

Excited about this, and just amazing, all the work here. Kudos to you all.

**Comment by Aspire1Inspire2** [comment]
Is it possible to integrate pandas with the @parallel decorator in ipyparallel, like the example that they have with numpy? 

http://ipyparallel.readthedocs.org/en/latest/multiengine.html#remote-function-decorators

I think theoretically speaking, even pandas does not support parallel computing by default, user can still refer to mpi4py for parallel computation. It's just some more coding time if one knows about MPI already.

**Comment by datnamer** [comment]
@Aspire1Inspire2  check out dask dataframe. https://jakevdp.github.io/blog/2015/08/14/out-of-core-dataframes-in-python/

**Comment by jreback** [comment]
let's just doc this to direct to dask

**Comment by heroxbd** [comment]
I tried out groupby and apply with pandas 0.17.1, and surprised to see the function is applied in parallel.  I am confused, is this feature already added and enabled by default?  I am not using dask.

**Comment by jreback** [comment]
@heroxbd well the GIL IS released during groupby operations. However there is not paralleism natively / inherent in groupby. Why do you think its in parallel?

**Comment by heroxbd** [comment]
The evidence is that all the loads of my CPUs grow to 100% when I call groupby apply.
`
    sph = tt.groupby('ev').apply(sphericity)
`

**Comment by jreback** [comment]
so if you do arithmetic ops inside ``sphericity`` (e.g. ``df + df``) or like. These defer to ``numexpr``, which utilizes multi-cores by default. Can you provide a sketch of this function?

These are performance things that pandas does w/o the user being specifically aware.

**Comment by heroxbd** [comment]
@jreback Ah-ha.  Here it is:
    pmtl, ql = event['pmt'].values, event['q'].values
    pl = pdir[pmtl] * ql[:,np.newaxis]
    s = np.sum(pl[:,:,np.newaxis] * pl[:,np.newaxis,:], axis=0)
    qs = np.sum(ql**2)
    eig = np.sort(np.linalg.eigvals(s/qs))
    return pd.Series({'S': (eig[0] + eig[1])*1.5, 'A': eig[0]*1.5}) # sphericity, aplanarity

**Comment by heroxbd** [comment]
Sorry for the noise. The parallel execution come from OpenMP used by OpenBLAS, which in turn is used by NumPy.

**Comment by jreback** [comment]
closing this as ``dask`` is the most appropriate for this. Certainly could add some usage within pandas to actually use dask, but those are separate issues (e.g. on a big-enough groupby (number of groups) deferring to dask is a good thing)

**Comment by dragoljub** [comment]
Was there a reason that pandas releasing the GIL in group by operations to only allows separate group by and apply operations to happen concurrently rather than the computing independent group aggregations in parallel?

**Comment by mrocklin** [comment]
Once you have GIL-releasing groupby operations then other developers can use Pandas in parallel.  It's actually quite tricky to intelligently write down the algorithms to handle all groupbys.  I think that if someone wants to do this for Pandas that'd be great.  It's a pretty serious undertaking though.

To do this with dask.dataframe

```
$ conda install dask
or 
$ pip install dask[dataframe]
```

```python
import pandas as pd
df = ... your pandas code ...

import dask.dataframe as dd
df = dd.from_pandas(df, npartitions=20)
dd_result = df.groupby(df.A).B.mean()  # or whatever
pd_result = dd_result.compute()  # uses as many threads as you have logical cores
```

Of course, this doesn't work on all groupbys (as mentioned before, this is incredibly difficult to do generally) but does work in many common cases.

**Comment by dragoljub** [comment]
@mrocklin thanks for the tip. How long would ```df = dd.from_pandas(df, npartitions=20)``` take on a dataframe with say 10M-100M rows? Is there a copy involved? Does dask support categorical columns?

**Comment by mrocklin** [comment]
There is a single copy involved (just to other pandas dataframes).  We're effectively just doing `dfs = df.iloc[i: i + blocksize] for i in range(0, len(df), blocksize)]`.  We're also doing a sort on the index ahead of time which, if your algorithms don't require it, you can turn off with `sort=False` (though some groupbys, particularly groupby.applys will definitely appreciate a sorted index.)

Generally the `from_pandas` call is cheap relative to groupby calls.  Copying data in memory generally runs fairly quickly.

**Comment by jreback** [comment]
@dragoljub 

In particular YMMV depending on such things as the number of groups that you are dealing and how  you partition

Small number of groups
```
In [7]: N = 10000000

In [8]: ngroups = 100

In [9]: df = pd.DataFrame({'A' : np.random.randn(N), 'B' : np.random.randint(0,ngroups,size=N)})

In [10]: %timeit df.groupby('B').A.mean()
10 loops, best of 3: 161 ms per loop
In [15]: ddf = dd.from_pandas(df, npartitions=8)

In [16]: %timeit ddf.groupby(ddf.B).mean().compute()
1 loop, best of 3: 223 ms per loop
```

Larger number of groups
```
In [17]: ngroups = 10000

In [18]: df = pd.DataFrame({'A' : np.random.randn(N), 'B' : np.random.randint(0,ngroups,size=N)})

In [19]: %timeit df.groupby('B').A.mean()
1 loop, best of 3: 591 ms per loop

In [21]: %timeit ddf.groupby(ddf.B).mean().compute()
1 loop, best of 3: 323 ms per loop
```

Can do even better if actually use our index
```
In [32]: ddf = dd.from_pandas(df.set_index('B'), npartitions=8)

In [33]: %timeit ddf.groupby(ddf.index).mean().compute()
1 loop, best of 3: 215 ms per loop
```

Note that these are pretty naive timings. This is a *single* computation that is split into embarassingly parallel tasks. Generally you would use dask for multiple steps in a computation. If you data *doesn't* fit in memory that dask can often help a lot more.

**Comment by rbiswas4** [comment]
In an embarrassingly parallel calculation, I create many dataframes which must be dumped to disc (This is the desired output of the program). I tried doing the computation and the dumping (to hdf5) in parallel  using joblib. And I run into HDFWrite trouble. Note, that at this point, I do not worry so much about performance of the parallel write. 

An example which demonstrates the problem is in https://github.com/rbiswas4/ParallelHDFWriteProblem

The program `demo_problem.py` includes a function called worker(i) which creates a very simple dataframe based on the input i and appends that to an hdf file. I do this twice once in serial, and once in parallel using joblib. 

What I find is that the serial case always works. But the parallel case is not reproducible: Sometimes it works without a problem and sometimes it crashes. The two log files https://github.com/rbiswas4/ParallelHDFWriteProblem/blob/master/demo.log.worked 
and 
https://github.com/rbiswas4/ParallelHDFWriteProblem/blob/master/demo.log_problem
are two cases when it worked and did not work respective. 

Is there a better way to write to hdf files in a parallel way from pandas that I should use? Is this a question for other fora like joblib and pyTables?

**Comment by shoyer** [comment]
@rbiswas4 If you want to dump a bunch of data to disk in parallel, the easiest thing to do is to create a separate HDF5 file for each process. Your approach is certainly going to result in corrupted data -- see the pytables FAQ for more details (http://www.pytables.org/FAQ.html#can-pytables-be-used-in-concurrent-access-scenarios). You might also be interested in dask.

**Comment by rbiswas4** [comment]
Writing to separate files is something I have to avoid because I might end up creating too many files (inode limits).  I suppose this means one of the following:
- I should not be looking at hdf5 as a possible output format, but use a database.
- Use a smaller number of partitions and hence files created. This is still quite inelegant
- consider some methodology by which I can write out the dataframes through a separate process (I am not sure how to split things)

It seems the first is the best bet.  And, yes I intend to see whether I should use `dask`! 

Thanks @shoyer

**Comment by heroxbd** [comment]
@rbiswas4, if you can write out raw hdf5 (via h5py) instead of pytables, please have a look at SWMR,

  https://www.hdfgroup.org/HDF5/docNewFeatures/NewFeaturesSwmrDocs.html

available in hdf5-1.10.
